pkg/controller/dmcluster/dm_cluster_controller.go

```diff
@@ -0,0 +1,364 @@
+	dcLister listers.DMClusterLister
+	// dcListerSynced returns true if the dmclusters shared informer has synced at least once
+	dcListerSynced cache.InformerSynced
+	// setLister is able to list/get stateful sets from a shared informer's store
+	setLister appslisters.StatefulSetLister
+	// setListerSynced returns true if the statefulset shared informer has synced at least once
+	setListerSynced cache.InformerSynced
+	// dmclusters that need to be synced.
+	queue workqueue.RateLimitingInterface
+}
+
+// NewController creates a dm controller.
+func NewController(
+	kubeCli kubernetes.Interface,
+	cli versioned.Interface,
+	genericCli client.Client,
+	informerFactory informers.SharedInformerFactory,
+	kubeInformerFactory kubeinformers.SharedInformerFactory,
+	autoFailover bool,
+) *Controller {
+	eventBroadcaster := record.NewBroadcasterWithCorrelatorOptions(record.CorrelatorOptions{QPS: 1})
+	eventBroadcaster.StartLogging(klog.V(2).Infof)
+	eventBroadcaster.StartRecordingToSink(&eventv1.EventSinkImpl{
+		Interface: eventv1.New(kubeCli.CoreV1().RESTClient()).Events("")})
+	recorder := eventBroadcaster.NewRecorder(v1alpha1.Scheme, corev1.EventSource{Component: "tidb-controller-manager"})
+
+	dcInformer := informerFactory.Pingcap().V1alpha1().DMClusters()
+	setInformer := kubeInformerFactory.Apps().V1().StatefulSets()
+	svcInformer := kubeInformerFactory.Core().V1().Services()
+	epsInformer := kubeInformerFactory.Core().V1().Endpoints()
+	pvcInformer := kubeInformerFactory.Core().V1().PersistentVolumeClaims()
+	pvInformer := kubeInformerFactory.Core().V1().PersistentVolumes()
+	podInformer := kubeInformerFactory.Core().V1().Pods()
+	//nodeInformer := kubeInformerFactory.Core().V1().Nodes()
+	//secretInformer := kubeInformerFactory.Core().V1().Secrets()
+
+	dcControl := controller.NewRealDMClusterControl(cli, dcInformer.Lister(), recorder)
+	masterControl := dmapi.NewDefaultMasterControl(kubeCli)
+	setControl := controller.NewRealStatefuSetControl(kubeCli, setInformer.Lister(), recorder)
+	svcControl := controller.NewRealServiceControl(kubeCli, svcInformer.Lister(), recorder)
+	pvControl := controller.NewRealPVControl(kubeCli, pvcInformer.Lister(), pvInformer.Lister(), recorder)
+	pvcControl := controller.NewRealPVCControl(kubeCli, recorder, pvcInformer.Lister())
+	// podControl := controller.NewRealPodControl(kubeCli, masterControl, podInformer.Lister(), recorder)
+	typedControl := controller.NewTypedControl(controller.NewRealGenericControl(genericCli, recorder))
+
+	dcc := &Controller{
+		kubeClient: kubeCli,
+		cli:        cli,
+		control: NewDefaultDMClusterControl(
+			dcControl,
+			mm.NewMasterMemberManager(
+				masterControl,
+				setControl,
+				svcControl,
+				// podControl,
+				typedControl,
+				setInformer.Lister(),
+				svcInformer.Lister(),
+				podInformer.Lister(),
+				epsInformer.Lister(),
+				pvcInformer.Lister(),
+			),
+			mm.NewWorkerMemberManager(
+				masterControl,
```

Comment:
`masterControl ` here?

Reply:
worker controller need `masterControl` to generate `masterClient` and then update worker status through master interface